Apps/Common/src/Services/AccessTokenService.cs

```diff
@@ -0,0 +1,172 @@
+    using HealthGateway.Common.ErrorHandling;
+    using HealthGateway.Common.Models.PHSA;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+
+    /// <inheritdoc/>
+    public class AccessTokenService : IAccessTokenService
+    {
+        /// <summary>
+        /// Configuration section to retrieve cache information from.
+        /// </summary>
+        private const string CacheConfigSectionName = "AuthCache";
+
+        /// <summary>
+        /// The generic cache domain to store access token against.
+        /// </summary>
+        private const string TokenSwapCacheDomain = "TokenSwap";
+
+        private readonly IAuthenticationDelegate authenticationDelegate;
+        private readonly ICacheProvider cacheProvider;
+        private readonly ILogger<AccessTokenService> logger;
+        private readonly int tokenCacheMinutes;
+        private readonly ITokenSwapDelegate tokenSwapDelegate;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AccessTokenService"/> class.
+        /// </summary>
+        /// <param name="logger">The logger to use.</param>
+        /// <param name="tokenSwapDelegate">The delegate responsible for swapping tokens with PHSA.</param>
+        /// <param name="cacheProvider">The cache provider responsible for caching.</param>
+        /// <param name="authenticationDelegate">The delegate responsible for fetching tokens.</param>
+        /// <param name="configuration">The configuration to use.</param>
+        public AccessTokenService(
+            ILogger<AccessTokenService> logger,
+            ITokenSwapDelegate tokenSwapDelegate,
+            ICacheProvider cacheProvider,
+            IAuthenticationDelegate authenticationDelegate,
+            IConfiguration configuration)
+        {
+            this.logger = logger;
+            this.tokenSwapDelegate = tokenSwapDelegate;
+            this.cacheProvider = cacheProvider;
+            this.authenticationDelegate = authenticationDelegate;
+
+            IConfigurationSection? configSection = configuration.GetSection(CacheConfigSectionName);
+            this.tokenCacheMinutes = configSection?.GetValue("TokenCacheExpireMinutes", 0) ?? 0;
+        }
+
+        private static ActivitySource Source { get; } = new(nameof(AccessTokenService));
+
+        /// <inheritdoc/>
+        public async Task<RequestResult<TokenSwapResponse>> GetPhsaAccessToken(string hdid)
+        {
+            using Activity? activity = Source.StartActivity();
+            RequestResult<TokenSwapResponse> requestResult = new();
+            TokenSwapResponse? cachedAccessToken = this.GetFromCache(hdid);
+
+            if (cachedAccessToken == null)
+            {
+                string? accessToken = this.authenticationDelegate.FetchAuthenticatedUserToken();
+
+                if (accessToken != null)
+                {
+                    requestResult = await this.SwapToken(hdid, accessToken).ConfigureAwait(true);
```

Comment:
It doesn't look like you cache the response

Reply:
See line 166. I broke out the calls instead of having one massive call.